File: cpp/744. Find Smallest Letter Greater Than Target.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
// Time: O(logn), Space: O(1)
public:
    char nextGreatestLetter(vector<char>& letters, char target) {
        int l = 0, r = letters.size();
        while(l < r) {
            int m = l + (r-l)/2;
            if(letters[m] > target) r = m;
            else l = m + 1;
        }
        if(l == letters.size()) return letters[0];
        return letters[l];
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vector<char> letters; char target; char ans;
    
    solution = Solution();
    letters = {'c', 'f', 'j'}, target = 'a';
    ans = solution.nextGreatestLetter(letters, target);
    assert(ans=='c');
}

File: python/744-Find Smallest Letter Greater Than Target.py
from typing import List

class Solution:
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        "Sequential Search: O(n), Space: O(1)"
        for letter in letters:
            if letter > target: return letter
        return letters[0]
    def nextGreatestLetter(self, letters: List[str], target: str) -> str:
        "Binary Search: O(logn), Space: O(1)"
        l, r = 0, len(letters)
        while l < r:
            m = l + (r-l)//2
            if letters[m] > target: r = m
            else: l = m + 1
        if l == len(letters): return letters[0]
        return letters[l]

letters = ["c", "f", "j"]
target = "a"
result = Solution().nextGreatestLetter(letters, target)
assert result == 'c'

letters = ["c", "f", "j"]
target = "c"
result = Solution().nextGreatestLetter(letters, target)
assert result == 'f'

letters = ["c", "f", "j"]
target = "d"
result = Solution().nextGreatestLetter(letters, target)
assert result == 'f'

letters = ["c", "f", "j"]
target = "g"
result = Solution().nextGreatestLetter(letters, target)
assert result == 'j'

letters = ["c", "f", "j"]
target = "j"
result = Solution().nextGreatestLetter(letters, target)
assert result == 'c'

letters = ["c", "f", "j"]
target = "k"
result = Solution().nextGreatestLetter(letters, target)
assert result == 'c'
